docs/proposals/cloud-edge-custom-message-design.md

```diff
@@ -0,0 +1,429 @@
+
+    2. **eventbus**: It is a source ruleEndpoint to send data to the cloud, or a target ruleEndpoint to receive messages from the cloud. 
+
+    3. **servicebus**:  a rest api on edge node. It is a target ruleEndpoint to receive messages delivered from cloud.
+
+* Rule defines how messages were delivered from source ruleendpoint to target ruleendpoint. It contained 3 rule types.
+
+   1. rest->eventbus: users' app calls rest api in cloud to send messages, finally  messages were sent to mqttbroker in edge.
+   2. eventbus->rest: users publish messages to mqttbroker in edge, finally messages were sent to rest api in cloud.
+   3. rest->servicebus: users' app calls rest api in cloud to send messages, finally  messages were sent to user's app in edge.
+
+### RuleEndpoint and Rule CRD design
+* RuleEndpoint
+```
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: ruleendpoints.rules.kubeedge.io
+spec:
+  group: rules.kubeedge.io
+  versions:
+    - name: v1
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
+          type: object
+          properties:
+            spec:
+              type: object
+              properties:
+                ruleEndpointType:
+                  description: 'ruleEndpointType is a string value representing rule-endpoint type. its value is one of rest/eventbus/servicebus.'
+                  type: string
+                properties:
+                  description: 'properties is not required except for servicebus rule-endpoint type. It is a map value representing rule-endpoint properties.
+                    when ruleEndpointType is servicebus, its value is {"service_port":"8080"}.'
+                  type: object
+                  additionalProperties:
+                    type: string
+              required:
+                - ruleEndpointType
+  scope: Namespaced
+  names:
+    plural: ruleendpoints
+    singular: ruleendpoint
+    kind: RuleEndpoint
+    shortNames:
+      - re
+
+```
+
+* Rule 
+```
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: rules.rules.kubeedge.io
+spec:
+  group: rules.kubeedge.io
+  versions:
+    - name: v1
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
+          type: object
+          properties:
+            spec:
+              type: object
+              properties:
+                source:
+                  description: 'source is a string value representing where the messages come from. Its
+                    value is the same with rule-endpoint name. For example, my-rest or my-eventbus.'
+                  type: string
+                sourceResource:
+                  description: 'sourceResource is a map representing the resource info of source. For rest
+                    rule-endpoint type its value is {"path":"/a/b"}. For eventbus rule-endpoint type its
```

Comment:
Does the path value need to be exactly matched, or prefix matched?

Reply:
prefix matched